src/main/java/org/opensearch/ad/ml/CheckpointDao.java

```diff
@@ -120,79 +139,96 @@
+        this.deserializeRCFBufferPool = deserializeRCFBufferPool;
+        this.deserializeRCFBufferSize = deserializeRCFBufferSize;
     }
 
-    /**
-     * Puts a model checkpoint in the storage.
-     *
-     * @deprecated use putModelCheckpoint with listener instead
-     *
-     * @param modelId Id of the model
-     * @param modelCheckpoint Checkpoint data of the model
-     */
-    @Deprecated
-    public void putModelCheckpoint(String modelId, String modelCheckpoint) {
-        Map<String, Object> source = new HashMap<>();
-        source.put(FIELD_MODEL, modelCheckpoint);
-        source.put(TIMESTAMP, ZonedDateTime.now(ZoneOffset.UTC));
+    private void saveModelCheckpointSync(Map<String, Object> source, String modelId) {
+        clientUtil.<IndexRequest, IndexResponse>timedRequest(new IndexRequest(indexName).id(modelId).source(source), logger, client::index);
+    }
 
+    private void putModelCheckpoint(String modelId, Map<String, Object> source, ActionListener<Void> listener) {
         if (indexUtil.doesCheckpointIndexExist()) {
-            saveModelCheckpointSync(source, modelId);
+            saveModelCheckpointAsync(source, modelId, listener);
         } else {
-            onCheckpointNotExist(source, modelId, false, null);
+            onCheckpointNotExist(source, modelId, true, listener);
         }
     }
 
-    private void saveModelCheckpointSync(Map<String, Object> source, String modelId) {
-        clientUtil.<IndexRequest, IndexResponse>timedRequest(new IndexRequest(indexName).id(modelId).source(source), logger, client::index);
+    /**
+     * Puts a rcf model checkpoint in the storage.
+     *
+     * @param modelId id of the model
+     * @param forest the rcf model
+     * @param listener onResponse is called with null when the operation is completed
+     */
+    public void putRCFCheckpoint(String modelId, RandomCutForest forest, ActionListener<Void> listener) {
+        Map<String, Object> source = new HashMap<>();
+        String modelCheckpoint = rcfModelToCheckpoint(forest);
+        source.put(FIELD_MODELV2, modelCheckpoint);
+        source.put(TIMESTAMP, ZonedDateTime.now(ZoneOffset.UTC));
+        putModelCheckpoint(modelId, source, listener);
     }
 
     /**
-     * Puts a model checkpoint in the storage.
+     * Puts a thresholding model checkpoint in the storage.
      *
      * @param modelId id of the model
-     * @param modelCheckpoint checkpoint of the model
+     * @param threshold the thresholding model
      * @param listener onResponse is called with null when the operation is completed
      */
-    public void putModelCheckpoint(String modelId, String modelCheckpoint, ActionListener<Void> listener) {
+    public void putThresholdCheckpoint(String modelId, ThresholdingModel threshold, ActionListener<Void> listener) {
+        String modelCheckpoint = AccessController.doPrivileged((PrivilegedAction<String>) () -> gson.toJson(threshold));
```

Comment:
Why does threshold mode put to FIELD_MODEL instead of FIELD_MODELV2?

Reply:
Because I don't want to deal with backward compatibility when it is not needed.  Currently, threshold model is in FIELD_MODEL.